tiledb/sm/query/dense_tiler.h

```diff
@@ -0,0 +1,354 @@
+   * attribute.
+   *
+   * @param id The id of the tile within the subarray to be retrieved.
+   *     The id is serialied in the tile order of the array domain.
+   * @param name The name of the attribute.
+   * @param tile The tile to be retrieved. This needs to
+   *     be preallocated and initialized before passed to the function.
+   * @return Status
+   */
+  Status get_tile(uint64_t id, const std::string& name, Tile* tile);
+
+  /**
+   * Retrieves the validity tile with the input id and for the input
+   * attribute.
+   *
+   * @param id The id of the tile within the subarray to be retrieved.
+   *     The id is serialied in the tile order of the array domain.
+   * @param name The name of the attribute.
+   * @param tile The tile to be retrieved. This needs to
+   *     be preallocated and initialized before passed to the function.
+   * @return Status
+   */
+  Status get_tile_null(uint64_t id, const std::string& name, Tile* tile) const;
+
+  /**
+   * Retrieves the var-sized tile with the input id and for the input
+   * attribute.
+   *
+   * @param id The id of the tile to be retrieved. The id is serialied in the
+   *     tile order of the array domain.
+   * @param name The name of the attribute.
+   * @param tile_off The tile with the offsets to be retrieved. This needs to
+   *     be preallocated and initialized before passed to the function.
+   * @param tile_val The tile with the values to be retrieves. This needs to
+   *     be preallocated and initialized before passed to the function.
+   * @return Status
+   */
+  Status get_tile_var(
+      uint64_t id,
+      const std::string& name,
+      Tile* tile_off,
+      Tile* tile_var) const;
+
+  /**
+   * Returns the number of tiles to be created. This is equal
+   * to the number of tiles intersecting the subarray.
+   */
+  uint64_t tile_num() const;
+
+  /**
+   * Returns the number of elements to "jump" in the tile when a dimension
+   * index changes.
+   */
+  const std::vector<uint64_t>& tile_strides_el() const;
+
+  /**
+   * Returns the number of elements to "jump" in the subarray when
+   * a dimension index changes.
+   */
+  const std::vector<uint64_t>& sub_strides_el() const;
+
+  /**
+   * Returns the strides used in calculating tile coordinates for the
+   * subarray when given a serial id.
+   */
+  const std::vector<uint64_t>& sub_tile_coord_strides() const;
+
+  /** Returns the coordinates of the first tile intersecting the subarray. */
+  const std::vector<uint64_t>& first_sub_tile_coords() const;
+
+  /** Returns the dense tiler stats. */
+  const stats::Stats& stats() const;
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /** The array schema. */
+  const ArraySchema* array_schema_;
+
+  /** The input buffers, from which the tiles will be produced. */
+  std::reference_wrapper<const std::unordered_map<std::string, QueryBuffer>>
```

Comment:
Similar to `array_schema_` and `subarray_`, let's use a raw pointer here as well instead of a reference wrapper.